openlane/state/design_format.py

```diff
@@ -40,141 +52,181 @@ class DesignFormatObject:
+        version="3.0.0",
+        action="once",
     )
+    def value(self) -> DesignFormat:
+        return self
 
-    JSON_HEADER: DesignFormatObject = DesignFormatObject(
-        "json_h",
-        "h.json",
-        "Design JSON Header File",
-    )
-    VERILOG_HEADER: DesignFormatObject = DesignFormatObject(
-        "vh",
-        "vh",
-        "Verilog Header",
-    )
+    def register(self):
+        self.__class__.factory.register(self)
 
     def __str__(self) -> str:
-        return self.value.id
+        return self.id
+
+    def __hash__(self):
+        return hash(self.id)
 
     @staticmethod
+    @deprecated(
+        "Use DesignFormat.factory.get",
+        version="3.0.0",
+        action="once",
+    )
     def by_id(id: str) -> Optional["DesignFormat"]:
-        return _designformat_by_id.get(id)
-
-
-_designformat_by_id: Dict[str, "DesignFormat"] = {
-    format.value.id: format for format in DesignFormat
-}
+        return DesignFormat.factory.get(id)
+
+    class DesignFormatFactory(object):
+        """
+        A factory singleton for DesignFormats, allowing them to be registered
+        and then retrieved by a string name.
+
+        See https://en.wikipedia.org/wiki/Factory_(object-oriented_programming) for
+        a primer.
+        """
+
+        _registry: ClassVar[Dict[str, DesignFormat]] = {}
+
+        @classmethod
+        def register(Self, df: DesignFormat) -> DesignFormat:
+            """
+            Adds a DesignFormat to the registry using its
+            :attr:`DesignFormat.id`, :attr:`DesignFormat.name` and
+            :attr:`DesignFormat.alts` attributes.
+            """
+            Self._registry[df.id] = df
+            Self._registry[df.name] = df
+            for alt in df.alts:
+                Self._registry[alt] = df
+            return df
+
+        @classmethod
+        def get(Self, name: str) -> Optional[DesignFormat]:
+            """
+            Retrieves a DesignFormat type from the registry using a lookup
+            string.
+
+            :param name: The registered name of the Step. Case-insensitive.
+            """
+            return Self._registry.get(name)
+
+        @classmethod
+        def list(Self) -> List[str]:
+            """
+            :returns: A list of IDs of all registered DesignFormat.
+            """
+            return [cls.id for cls in Self._registry.values()]
+
+    factory: ClassVar = DesignFormatFactory
+
+
+# Common Design Formats
+DesignFormat(
+    "nl",
+    "nl.v",
+    "Verilog Netlist",
+    "NETLIST",
+).register()
+
+DesignFormat(
+    "pnl",
+    "pnl.v",
+    "Powered Verilog Netlist",
+    "POWERED_NETLIST",
+).register()
+
+DesignFormat(
+    "sdf_pnl",
+    "sdf_pnl.v",
+    "Powered Verilog Netlist for SDF Simulation (No Fills)",
+    "POWERED_NETLIST_SDF_FRIENDLY",
+    folder_override="pnl",
+    alts=["pnl-sdf"],
+).register()
+
+DesignFormat(
+    "npc_pnl",
```

Comment:
would rather have it named differently now. Also, I think `npc` is bad and ambiguous - especially after the rename.

Reply:
Aligned